File: src/core/Stocks/OwnedStock.cs
using System;
using System.Collections.Generic;
using System.Linq;
using core.Shared;

namespace core.Stocks
{
    public class OwnedStock : Aggregate<OwnedStockState>
    {
        public OwnedStock(IEnumerable<AggregateEvent> events) : base(events)
        {
        }

        public OwnedStock(Ticker ticker, Guid userId)
        {
            if (userId == Guid.Empty)
            {
                throw new InvalidOperationException("Missing user id");
            }

            Apply(new TickerObtained(Guid.NewGuid(), Guid.NewGuid(), DateTimeOffset.UtcNow, ticker.Value, userId));
        }

        public void Purchase(decimal numberOfShares, decimal price, DateTimeOffset date, string notes = null, decimal? stopPrice = null)
        {
            if (price <= 0)
            {
                throw new InvalidOperationException("Price cannot be negative or zero");
            }

            if (date == DateTime.MinValue)
            {
                throw new InvalidOperationException("Purchase date not specified");
            }

            Apply(
                new StockPurchased_v2(
                    Guid.NewGuid(),
                    State.Id,
                    date,
                    State.UserId,
                    State.Ticker.Value,
                    numberOfShares,
                    price,
                    notes,
                    stopPrice
                )
            );
        }

        public void DeleteStop()
        {
            if (State.OpenPosition == null)
            {
                throw new InvalidOperationException("No open position to delete stop for");
            }

            Apply(
                new StopDeleted(
                    Guid.NewGuid(),
                    State.Id,
                    DateTimeOffset.UtcNow,
                    State.UserId,
                    State.Ticker.Value
                )
            );
        }

        public void SetRiskAmount(decimal riskAmount, int positionId)
        {
            var position = State.GetPosition(positionId);
            if (position == null)
            {
                throw new InvalidOperationException("Unable to find position with id " + positionId);
            }

            if (riskAmount < 0)
            {
                throw new InvalidOperationException("Risk amount cannot be negative");
            }

            Apply(new PositionRiskAmountSet(Guid.NewGuid(), State.Id, DateTimeOffset.UtcNow, State.UserId, positionId, riskAmount));
        }

        public bool SetStop(decimal stopPrice)
        {
            if (State.OpenPosition == null)
            {
                throw new InvalidOperationException("No open position to set stop on");
            }

            if (stopPrice < 0)
            {
                throw new InvalidOperationException("Stop price cannot be negative");
            }

            if (stopPrice == State.OpenPosition.StopPrice)
            {
                return false;
            }

            Apply(
                new StopPriceSet(Guid.NewGuid(), State.Id, DateTimeOffset.UtcNow, State.UserId, State.Ticker.Value, stopPrice)
            );

            return true;
        }

        public bool AddNotes(string notes)
        {
            if (State.OpenPosition == null)
            {
                throw new InvalidOperationException("No open position to add notes to");
            }

            if (string.IsNullOrEmpty(notes))
            {
                return false;
            }

            if (State.OpenPosition.Notes.Contains(notes))
            {
                return false;
            }

            Apply(
                new NotesAdded(
                    Guid.NewGuid(),
                    State.Id,
                    DateTimeOffset.UtcNow,
                    State.UserId,
                    State.OpenPosition.PositionId,
                    notes
                )
            );

            return true;
        }

        public void DeleteTransaction(Guid transactionId)
        {
            if (!State.BuyOrSell.Any(t => t.Id == transactionId))
            {
                throw new InvalidOperationException("Unable to find transcation to delete using id " + transactionId);
            }

            Apply(
                new StockTransactionDeleted(
                    Guid.NewGuid(),
                    State.Id,
                    transactionId,
                    DateTimeOffset.UtcNow
                    
                )
            );
        }

        public void Delete()
        {
            Apply(
                new StockDeleted(
                    Guid.NewGuid(),
                    State.Id,
                    DateTimeOffset.UtcNow
                )
            );
        }

        public void Sell(decimal numberOfShares, decimal price, DateTimeOffset date, string notes)
        {
            if (State.OpenPosition == null)
            {
                throw new InvalidOperationException("No open position to sell");
            }
            
            if (State.OpenPosition.NumberOfShares < numberOfShares)
            {
                throw new InvalidOperationException("Cannot sell more shares than owned");
            }

            if (price < 0)
            {
                throw new InvalidOperationException("Price cannot be negative or zero");
            }

            Apply(
                new StockSold(
                    Guid.NewGuid(),
                    State.Id,
                    date,
                    State.UserId,
                    State.Ticker.Value,
                    numberOfShares,
                    price,
                    notes)
            );
        }

        public bool AssignGrade(int positionId, TradeGrade grade, string note)
        {
            var position = State.GetPosition(positionId);
            if (position == null)
            {
                throw new InvalidOperationException("Unable to find position with id " + positionId);
            }

            if (!position.IsClosed)
            {
                throw new InvalidOperationException("Cannot assign grade to an open position");
            }

            if (position.Grade.Equals(grade) && position.GradeNote == note)
            {
                return false;
            }

            Apply(
                new TradeGradeAssigned(
                    Guid.NewGuid(),
                    State.Id,
                    DateTimeOffset.UtcNow,
                    userId: State.UserId,
                    grade: grade.Value,
                    note: note,
                    positionId: positionId
                )
            );

            return true;
        }

        public bool DeletePosition(int positionId)
        {
            var position = State.GetPosition(positionId);
            if (position == null) // already deleted before, moving on
            {
                return false;
            }

            // we don't want to mess with closed positions, at leats for now
            if (position.IsClosed)
            {
                throw new InvalidOperationException("Cannot delete a closed position");
            }

            Apply(
                new PositionDeleted(
                    Guid.NewGuid(),
                    State.Id,
                    DateTimeOffset.UtcNow,
                    State.UserId,
                    positionId
                )
            );

            return true;
        }

        public bool SetPositionLabel(int positionId, string key, string value)
        {
            var position = State.GetPosition(positionId);
            if (position == null)
            {
                throw new InvalidOperationException("Unable to find position with id " + positionId);
            }

            if (string.IsNullOrEmpty(key))
            {
                throw new InvalidOperationException("Key cannot be empty");
            }

            if (string.IsNullOrEmpty(value))
            {
                throw new InvalidOperationException("Value cannot be empty");
            }

            if (position.ContainsLabel(key, value))
            {
                return false;
            }

            Apply(
                new PositionLabelSet(
                    Guid.NewGuid(),
                    State.Id,
                    DateTimeOffset.UtcNow,
                    State.UserId,
                    positionId,
                    key,
                    value
                )
            );

            return true;
        }

        public bool DeletePositionLabel(int positionId, string key)
        {
            var position = State.GetPosition(positionId);
            if (position == null)
            {
                throw new InvalidOperationException("Unable to find position with id " + positionId);
            }

            if (string.IsNullOrEmpty(key))
            {
                throw new InvalidOperationException("Key cannot be empty");
            }

            if (!position.ContainsLabel(key))
            {
                return false;
            }

            Apply(
                new PositionLabelDeleted(
                    Guid.NewGuid(),
                    State.Id,
                    DateTimeOffset.UtcNow,
                    State.UserId,
                    positionId,
                    key
                )
            );

            return true;
        }
    }
}

File: src/core2.fs/Stocks/OwnedStock.fs
namespace core.Stocks

open System
open System.Collections.Generic
open System.Linq
open core.Shared

type internal OwnedStock =
    inherit Aggregate<OwnedStockState>

    new (events: IEnumerable<AggregateEvent>) = { inherit Aggregate<OwnedStockState>(events) }

    new (ticker: Ticker, userId: Guid) as this =
        { inherit Aggregate<OwnedStockState>() }
        then
            if userId = Guid.Empty then
                raise (InvalidOperationException("Missing user id"))
            this.Apply(TickerObtained(Guid.NewGuid(), Guid.NewGuid(), DateTimeOffset.UtcNow, ticker.Value, userId))

    member this.Purchase(numberOfShares: decimal, price: decimal, date: DateTimeOffset, ?notes: string, ?stopPrice: decimal) =
        if price <= 0m then
            raise (InvalidOperationException("Price cannot be negative or zero"))

        if date.Subtract(DateTimeOffset.UtcNow).TotalHours >= 12.0 then
            raise (InvalidOperationException("Purchase date cannot be in the future"))

        if date < DateTimeOffset.UnixEpoch then
            raise (InvalidOperationException("Purchase date cannot be before 1970"))

        this.Apply(
            StockPurchased_v2(
                Guid.NewGuid(),
                this.State.Id,
                date,
                this.State.UserId,
                this.State.Ticker.Value,
                numberOfShares,
                price,
                defaultArg notes null,
                stopPrice
            )
        )

    member this.DeleteStop() =
        if isNull this.State.OpenPosition then
            raise (InvalidOperationException("No open position to delete stop for"))

        this.Apply(
            StopDeleted(
                Guid.NewGuid(),
                this.State.Id,
                DateTimeOffset.UtcNow,
                this.State.UserId,
                this.State.Ticker.Value
            )
        )

    member this.SetRiskAmount(riskAmount: decimal, positionId: int) =
        let position = this.State.GetPosition(positionId)
        if isNull position then
            raise (InvalidOperationException("Unable to find position with id " + string positionId))

        if riskAmount < 0m then
            raise (InvalidOperationException("Risk amount cannot be negative"))

        this.Apply(PositionRiskAmountSet(Guid.NewGuid(), this.State.Id, DateTimeOffset.UtcNow, this.State.UserId, positionId, riskAmount))

    member this.SetStop(stopPrice: decimal) =
        if isNull this.State.OpenPosition then
            raise (InvalidOperationException("No open position to set stop on"))

        if stopPrice < 0m then
            raise (InvalidOperationException("Stop price cannot be negative"))

        if Some stopPrice = this.State.OpenPosition.StopPrice then
            false
        else
            this.Apply(
                StopPriceSet(Guid.NewGuid(), this.State.Id, DateTimeOffset.UtcNow, this.State.UserId, this.State.Ticker.Value, stopPrice)
            )
            true

    member this.AddNotes(notes: string) =
        if isNull this.State.OpenPosition then
            raise (InvalidOperationException("No open position to add notes to"))

        if String.IsNullOrEmpty(notes) then
            false
        elif this.State.OpenPosition.Notes.Contains(notes) then
            false
        else
            this.Apply(
                NotesAdded(
                    Guid.NewGuid(),
                    this.State.Id,
                    DateTimeOffset.UtcNow,
                    this.State.UserId,
                    this.State.OpenPosition.PositionId,
                    notes
                )
            )
            true

    member this.DeleteTransaction(transactionId: Guid) =
        if this.State.BuyOrSell.All(fun t -> t.Id <> transactionId) then
            raise (InvalidOperationException("Unable to find transaction to delete using id " + string transactionId))

        this.Apply(
            StockTransactionDeleted(
                Guid.NewGuid(),
                this.State.Id,
                transactionId,
                DateTimeOffset.UtcNow
            )
        )

    member this.Delete() =
        this.Apply(
            StockDeleted(
                Guid.NewGuid(),
                this.State.Id,
                DateTimeOffset.UtcNow
            )
        )

    member this.Sell(numberOfShares: decimal, price: decimal, date: DateTimeOffset, notes: string) =
        if isNull this.State.OpenPosition then
            raise (InvalidOperationException("No open position to sell"))

        if this.State.OpenPosition.NumberOfShares < numberOfShares then
            raise (InvalidOperationException("Cannot sell more shares than owned"))

        if price < 0m then
            raise (InvalidOperationException("Price cannot be negative or zero"))

        this.Apply(
            StockSold(
                Guid.NewGuid(),
                this.State.Id,
                date,
                this.State.UserId,
                this.State.Ticker.Value,
                numberOfShares,
                price,
                notes
            )
        )

    member this.AssignGrade(positionId: int, grade: TradeGrade, note: string) =
        let position = this.State.GetPosition(positionId)
        if isNull position then
            raise (InvalidOperationException("Unable to find position with id " + string positionId))

        if not position.IsClosed then
            raise (InvalidOperationException("Cannot assign grade to an open position"))

        if position.Grade = Some grade && position.GradeNote = note then
            false
        else
            this.Apply(
                TradeGradeAssigned(
                    Guid.NewGuid(),
                    this.State.Id,
                    DateTimeOffset.UtcNow,
                    userId = this.State.UserId,
                    grade = grade.Value,
                    note = note,
                    positionId = positionId
                )
            )
            true

    member this.DeletePosition(positionId: int) =
        let position = this.State.GetPosition(positionId)
        if isNull position then
            false
        elif position.IsClosed then
            raise (InvalidOperationException("Cannot delete a closed position"))
        else
            this.Apply(
                PositionDeleted(
                    Guid.NewGuid(),
                    this.State.Id,
                    DateTimeOffset.UtcNow,
                    this.State.UserId,
                    positionId
                )
            )
            true

    member this.SetPositionLabel(positionId: int, key: string, value: string) =
        let position = this.State.GetPosition(positionId)
        if isNull position then
            raise (InvalidOperationException("Unable to find position with id " + string positionId))

        if String.IsNullOrEmpty(key) then
            raise (InvalidOperationException("Key cannot be empty"))

        if String.IsNullOrEmpty(value) then
            raise (InvalidOperationException("Value cannot be empty"))

        if position.ContainsLabel(key, value) then
            false
        else
            this.Apply(
                PositionLabelSet(
                    Guid.NewGuid(),
                    this.State.Id,
                    DateTimeOffset.UtcNow,
                    this.State.UserId,
                    positionId,
                    key,
                    value
                )
            )
            true

    member this.DeletePositionLabel(positionId: int, key: string) =
        let position = this.State.GetPosition(positionId)
        if isNull position then
            raise (InvalidOperationException("Unable to find position with id " + string positionId))

        if String.IsNullOrEmpty(key) then
            raise (InvalidOperationException("Key cannot be empty"))

        if not (position.ContainsLabel(key)) then
            false
        else
            this.Apply(
                PositionLabelDeleted(
                    Guid.NewGuid(),
                    this.State.Id,
                    DateTimeOffset.UtcNow,
                    this.State.UserId,
                    positionId,
                    key
                )
            )
            true
